### ISSUE TITLE
Amount transferred to PayPal does not change after checkout recalculation (1769)

### ISSUE THREAD

**Issue by dennisnissle** [woocommerce/woocommerce-paypal-payments#1422]
### Describe the Bug

When dynamically changing amounts during checkout, e.g. on selecting a certain checkbox or after changing to a certain country, the total amount transferred to PayPal does not change (at least not within the PayPal modal shown to the customer before submitting the order). Seems like the original created order is not being updated after (re-)calculating cart totals.

### To Reproduce

1. Add the following snippet to your functions.php

```php
add_action( 'woocommerce_before_calculate_totals', function( $cart ) {
	$country = isset( $_POST['country'] ) ? wc_clean( $_POST['country'] ) : WC()->checkout()->get_posted_address_data( 'country' );

	if ( 'AT' === $country ) {
		foreach ( $cart->get_cart() as $cart_item_key => $values ) {
			$_product = apply_filters( 'woocommerce_cart_item_product', $values['data'], $values, $cart_item_key );
			$_product->set_price( 5 );
		}
	}
}, 15 );
```

2. Add a product to cart, go to checkout. Make sure that Austria is not your default country
3. Use the PayPal Smart Button which opens the payment modal (in case you are logged in make sure to choose another payment source and do not force a quick checkout)
4. Close the modal
5. Choose Austria as your billing country
6. See how the prices are updated within checkout/cart
7. Click the Smart Button again and see how the old total is presented in the modal

### Screenshots

<img width="1186" alt="Bildschirmfoto 2023-06-06 um 13 05 53" src="https://github.com/woocommerce/woocommerce-paypal-payments/assets/8328938/b2c87401-2002-4342-9229-8cfafe27747c">

1. Before cart adjustment

<img width="1232" alt="Bildschirmfoto 2023-06-06 um 13 06 26" src="https://github.com/woocommerce/woocommerce-paypal-payments/assets/8328938/6965f04a-6d9b-4bd7-860f-c98c7470f633">

2. After cart adjustment

### Expected Behavior

The original order is being after adjusting cart data, e.g. via the `woocommerce_after_calculate_totals` hook.

### Environment

 - WordPress Version: 6.2
 - WooCommerce Version: 7.7.2
 - Plugin Version: 2.1.0-rc1
 - Browser: Latest chrome, mac os x

**Comment by InpsydeNiklas** [comment]
Hi @dennisnissle 
Regardless of what is displayed in the PayPal popup window, the buyer will be charged the correct amount as this is taken from the WooCommerce order before the capture.

**Comment by dennisnissle** [comment]
Glad to hear that this is the case but it still is very misleading for the actual customer. I would not submit an order if the amount displayed by PayPal is plain wrong. I don't think that it's an _enhancement_ but a serious bug. 

Say someone opens the modal, closes it, applies a coupon and then reopens the modal. The modal still shows the old total amount.

**Comment by InpsydeNiklas** [comment]
> Say someone opens the modal, closes it, applies a coupon and then reopens the modal. The modal still shows the old total amount.

This is not the case when using native WooCommerce coupons:

https://github.com/woocommerce/woocommerce-paypal-payments/assets/70433191/b5d023bd-5cc6-4379-b438-d175be7bbe8e

Discounts made via third-party plugins or custom code may not always be detected correctly. But that would probably need to be handled individually for each plugin/use case.

**Comment by dennisnissle** [comment]
Maybe that issue is due to being logged in via PayPal? I've tested it with core WooCommerce coupons.

<img width="1325" alt="Bildschirmfoto 2023-06-07 um 11 22 53" src="https://github.com/woocommerce/woocommerce-paypal-payments/assets/8328938/4c74b244-5bb3-429a-9344-4aa89f600847">
<img width="1369" alt="Bildschirmfoto 2023-06-07 um 11 24 22" src="https://github.com/woocommerce/woocommerce-paypal-payments/assets/8328938/e27ff3cf-e372-4eb9-9e60-cf0ee8d97cb0">

**Comment by InpsydeNiklas** [comment]
Yes, this behavior seemingly can be reproduced only when the Vaulting feature is used from the Checkout page.

https://github.com/woocommerce/woocommerce-paypal-payments/assets/70433191/8f9461d8-2e6f-40d2-800c-877fe81eb8be

Specifically, the buyer has to have a PayPal account saved in the Vault. On the Checkout page, the buyer has to click the dropdown arrow on the PayPal button and choose to pay with another funding source/account for the modal to open.
In this case, the PayPal popup window may not display the updated totals (unless the page was reloaded after applying the coupon).
But when the buyer clicks the vaulted "Pay now" button as usual, there wouldn't even be a PayPal modal but only a one-click payment confirmation.

When not using the Vaulting feature, both use cases (your code snippet & WC coupons) update the total in the PayPal popup window as expected for me.
Just so you know, the Vaulting integration will soon be updated from Vault v2 to Vault v3, which should help resolve many different edge-case problems, and we'll keep track of this behavior.

**Comment by dennisnissle** [comment]
Thanks for the update. Is there any way to disable vaulting from the settings?

**Comment by InpsydeNiklas** [comment]
The Vaulting setting can be disabled from the Standard Payments tab: ![image](https://github.com/woocommerce/woocommerce-paypal-payments/assets/70433191/a5b3884f-e863-4dae-b150-7fbb6a5d5768)

In 2.0.5, the Vaulting setting is shared between PayPal and Advanced Card Processing, but the 2.1.0 update will split this up so the Vaulting feature can be individually enabled/disabled for PayPal or advanced card payments.
However, following that change vaulted PayPal accounts may still be available to use when you disable Vaulting for PayPal but not for Advanced Card Processing. So to entirely disable vaulted payment methods in 2.1.0-rc1 or newer, please disable the Vaulting setting in the Advanced Card Processing tab as well.
